### DISCUSSION TITLE
Porting Jainja JVM

### DISCUSSION THREAD

**Discussion by glegris** [mikaku/Fiwix#95] [Show and tell]
Thanks to @mikaku I managed to get the [Jainja JVM](https://jainja.thenesis.org/) (which I author) working partially.

Things that work well:
- Basic testing with java.lang
- Preemptive multithreading
- Graphics

Current difficulties: 
- Using a resolution higher than 320x200 (8 bpp) in VGA mode: SDL1 is the current graphics backend. Depths lower than 8 bpp are not supported and only the first mode is available (320x200). The best way to solve this problem will be to use the framebuffer.
- Mouse support
  A workaround to be able to use the mouse is to add `-chardev msmouse,id=chardev1 -device isa-serial,chardev=chardev1,id=serial1` to the QEMU command line and change some settings in /etc/vga/vgalib.conf:
   - Activate serial mouse: uncomment line `mdev /dev/ttyS0`
   - Disable mouse acceleration: `mouse_accel_type off`

@mikaku is working on the mouse driver and ioctl commands to manage framebuffer

![Jainja on Fiwix - Thinlet - 20240924](https://github.com/user-attachments/assets/7cd4eff6-e468-4eef-9ab0-9a9414ddf815)

![Jainja on Fiwix - Tetris MIDP- 20240924](https://github.com/user-attachments/assets/2b84402d-e6e7-48e9-8d4f-7fab2408bdff)

### Some details to help those who would like to port software

Since there is no network yet, I put the sources to compile in the [FiwixOS raw image](https://www.fiwix.org/downloads.html). For this I mounted a disk image using the loop device. I then compiled the code from Fiwix OS using gcc.

1.  Calculate the offset of the ext2 partition: 

```
fdisk -lu FiwixOS-3.3-i386.raw
```
   
> offset = 512 * Start

2. Create a loop device 
```
losetup -o <offset> /dev/loop0 FiwixOS-3.3-i386.raw
```
> Check if it's ok by listing loop devices: `losetup -l`

3. Mount the partition
```
mount /dev/loop0 /mnt/disk
```
4. Copy your code
```
cp mysoftware /mnt/disk/home
```

5. Unmount the partition 
```
umount /mnt/disk
```
```
losetup -d /dev/loop0
```
6. Start FiwixOS from the image
```
qemu-system-i386 -enable-kvm -m 2048M -drive format=raw,file=FiwixOS-3.3-i386.raw -chardev msmouse,id=chardev1 -device isa-serial,chardev=chardev1,id=serial1
```

**Comment by mikaku** [comment]
Just a small advice.

Since the default virtual memory split is 3GiB/1GiB (user/kernel), the Fiwix kernel will use only up to 1GiB of physical memory. So, defining more than 1GiB won't have any effect.

If you really need to have up to 2GiB of physical memory, you'll have to activate the virtual memory split of 2GiB/2GiB by recompiling the kernel with the option `CONFIG_VM_SPLIT22` enabled.
